### PR TITLE
Enhance the Linter to detect accidental upcasts

### PR DESCRIPTION
**Pull request by mgaudet**
In an extensible class hierarchy, it is important that each class in the OMR layer define the downcast function `self()`. Normally, missing this would cause a compilation failure.

However, if you have an extensible class hierarchy inheriting from another, there is another possible problem that until this PR, it hasn't  been detected: If the derived extensible class hierarchy never defines `self`, then `self()` will resolve to the parent hierarchy, making it an upcast rather than a downcast.

This PR fixes the only known issue of this type (`OMR::StaticSymbol`) , and enhances the linter to detect this kind of error

### CODE REVIEW

File: tools/compiler/OMRChecker/testing/input/bad/doubleExtension.cpp [hunk 1]
```diff
@@ -0,0 +1,58 @@
+/*******************************************************************************
+*
+* (c) Copyright IBM Corp. 2016, 2016
```
Comment (andrewcraik) [inline tools/compiler/OMRChecker/testing/input/bad/doubleExtension.cpp]: Please correct the copyright dates.

Comment (mgaudet) [inline tools/compiler/OMRChecker/testing/input/bad/doubleExtension.cpp]: Got it. Thanks :)

File: tools/compiler/OMRChecker/OMRChecker.hpp [hunk 2]
```diff
@@ -704,16 +719,30 @@ class OMRThisCheckingVisitor : public RecursiveASTVisitor<OMRThisCheckingVisitor
...
-
       Expr *receiver = call->getImplicitObjectArgument()->IgnoreParenImpCasts();
 
       if (receiver->isImplicitCXXThis()) {
+         trace("Reciever is implicit this");
          // Don't diagnose implicit this recievers that are calling self.
-         if (!isSelfCall(call)) {
+         if (isSelfCall(call)) { 
+            trace("callee is self()");
+            // Need still to verify that the self() call is to the right reciever! Otherwise
```
Comment (0xdaryl) [inline tools/compiler/OMRChecker/OMRChecker.hpp]: `i` before `e` except after `c`.

File: tools/compiler/OMRChecker/OMRChecker.hpp [hunk 3]
```diff
@@ -704,16 +719,30 @@ class OMRThisCheckingVisitor : public RecursiveASTVisitor<OMRThisCheckingVisitor
...
-      //TODO: Check that this member call expression is inside of an extensible class!
       if (!isExtensibleMemberCall(call)) {
+         trace("Call isn't an Extensible Member Call" << call);
          return true;
       }
-
       Expr *receiver = call->getImplicitObjectArgument()->IgnoreParenImpCasts();
 
       if (receiver->isImplicitCXXThis()) {
+         trace("Reciever is implicit this");
```
Comment (0xdaryl) [inline tools/compiler/OMRChecker/OMRChecker.hpp]: `i` before `e` except after `c`.

File: tools/compiler/OMRChecker/OMRChecker.hpp [hunk 4]
```diff
@@ -779,6 +809,77 @@ class OMRThisCheckingVisitor : public RecursiveASTVisitor<OMRThisCheckingVisitor
...
+
+   /**
+    * Return true iff the call is to self, **and** the self call is 
+    * within the correct extensible class string. 
+    *
+    * The correct call depends on what member function we are inside of.
+    *
+    */
+   bool isCorrectSelf(CXXMemberCallExpr *call) { 
+      // This is the callled method decl -- Should be self of one form or another, verified by isSelfCall.
```
Comment (0xdaryl) [inline tools/compiler/OMRChecker/OMRChecker.hpp]: `callled` -> `called`

**Comment by mgaudet** [comment]
Added https://github.com/eclipse/omr/pull/932/commits/c6e97649a97a269c29883b8a51bb843dda306ccf to correct spelling errors (there were more that were pre-existing).

**Pull request merged**